mlir/lib/Dialect/SPIRV/IR/ImageOps.cpp

```diff
@@ -29,10 +31,121 @@ static LogicalResult verifyImageOperands(Operation *imageOp,
+                            imageType.getDim()))
+      return imageOp->emitError("Lod only be used with an image type that has "
+                                "a dim operand of 1D, 2D, 3D, or Cube");
+
+    if (imageType.getSamplingInfo() != spirv::ImageSamplingInfo::SingleSampled)
+      return imageOp->emitError(
+          "Lod only be used with an image type that has a MS operand of 0");
+
+    ++index;
+  }
+
+  if (spirv::bitEnumContainsAny(attr.getValue(), spirv::ImageOperands::Grad)) {
+    if (!isa<spirv::ExplicitLodOpInterface>(imageOp))
+      return imageOp->emitError(
+          "Grad is only valid with explicit-lod instructions");
+
+    if (index + 2 > operands.size())
+      return imageOp->emitError(
+          "Grad operand requires 2 arguments (scalars or vectors)");
+
+    auto samplingOp = cast<spirv::SamplingOpInterface>(imageOp);
+    auto sampledImageType =
+        cast<spirv::SampledImageType>(samplingOp.getSampledImage().getType());
+    auto imageType = cast<spirv::ImageType>(sampledImageType.getImageType());
+
+    if (imageType.getSamplingInfo() != spirv::ImageSamplingInfo::SingleSampled)
+      return imageOp->emitError(
+          "Grad only be used with an image type that has a MS operand of 0");
+
+    int64_t numberOfComponents = 0;
+
+    auto coordVector =
+        dyn_cast<mlir::VectorType>(samplingOp.getCoordinate().getType());
+    if (coordVector) {
+      numberOfComponents = coordVector.getNumElements();
+      if (imageType.getArrayedInfo() == spirv::ImageArrayedInfo::Arrayed)
+        numberOfComponents -= 1;
+    } else
+      numberOfComponents = 1;
+
+    assert(numberOfComponents > 0);
+
+    auto dXVector = dyn_cast<mlir::VectorType>(operands[index].getType());
+    auto dYVector = dyn_cast<mlir::VectorType>(operands[index + 1].getType());
+    if (dXVector && dYVector) {
+      if (dXVector.getNumElements() != dYVector.getNumElements() ||
+          dXVector.getNumElements() != numberOfComponents)
+        return imageOp->emitError(
+            "number of components of each Grad argument must equal the number "
+            "of components in coordinate, minus the array layer component, if "
+            "present");
+
+      if (!isa<mlir::FloatType>(dXVector.getElementType()) ||
+          !isa<mlir::FloatType>(dYVector.getElementType()))
+        return imageOp->emitError(
+            "Grad arguments must be a vector of floating-point type");
+    } else if (isa<mlir::FloatType>(operands[index].getType()) &&
+               isa<mlir::FloatType>(operands[index + 1].getType())) {
+      if (numberOfComponents != 1)
+        return imageOp->emitError(
+            "number of components of each Grad argument must equal the number "
+            "of components in coordinate, minus the array layer component, if "
+            "present");
+    } else
+      return imageOp->emitError(
+          "Grad arguments must be a scalar or vector of floating-point type");
+
```

Comment:
```suggestion
    } else {
      return imageOp->emitError(
          "Grad arguments must be a scalar or vector of floating-point type");
   }
```

If you use braces in one path, make sure to also add it to all the other ones